docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,668 @@
+
+Having precise types for variants, we have two options:
+
+ 1. Keep the current encoding, that is, a variant is encoded as a super-record
+ that contains all possible fields of the member records. The type checker
+ will guarantee that we do not access the fields of the super-record that are
+ not present in the actual record type.
+
+ 1. Implement the
+ [suggestion by Shon Feder](https://github.com/informalsystems/apalache/discussions/789#discussioncomment-1592118).
+
+Although Option 2 looks nicer, we prefer keeping Option 1. The reason is that
+the current implementation introduces the minimal number of constraints
+by mashing all possible fields into a super-record. The alternative solution
+(option 2) would introduce additional constraints, when the spec requires us
+to extract an element from a set.
+
+Recall the example with a set of messages:
+
+```tla
+VARIABLES
+  (*
+    @typeAlias: MESSAGE =
+       [ tag: "1a", bal: Int ]
+     | [ tag: "2a", bal: Int, val: Int ]);
+    @type: Set(MESSAGE);
+   *)  
+  msgs
+```  
+
+Consider an existential quantifier over the variable `msgs`:
+
+```tla
+Next ==
+  \E m \in msgs:
+    P
+```
+
+In the current encoding, `m` is a super-record that contains three fields:
+`tag`, `bal`, and `val`, even if some of these fields are not required by the
+actual type of `m`. In the alternative encoding, `m` is a tuple `tup`,
+which equals to one of the following tuples, depending on the value of the
+field `m.tag`:
+
+```
+tup = IF m.tag = "1a"
+      THEN << { [ tag |-> "1a", bal |-> b ] }, {} >>
+      ELSE << {}, [ tag |-> "2a", bal |-> b, val |-> v ] }, {} >>
+```
+
+Since it is impossible to statically compute the actual type of `m`, the
+rewriting rules would have to replicate the structure of both elements of the
+tuple `tup`. This would lead to a blow-up in the number of constraints.
+
+### 5.4. Additional requirements
+
+Given the decisions in Section 5.3, we additionally require that all records in
+a variant type have compatible field types. In more detail, if a variant type
+contains two record types `[ tag: "A", value: a, ... ]` and `[ tag: "B", value:
+b, ...]`, then the types `a` and `b` must be unifiable.  In practice, this
+often implies that `a` and `b` are simply the same type.
+
+### 5.5. The Variants module
+
+We introduce a new module that is called `Variants.tla`. It contains the
+operators `Variant`, `FilterByTag`, `MatchTag`, and `MatchOnly`. This module
+will be distributed with Apalache. As is custom in the TLA+ community, the
+users should be also able to copy `Variants.tla` next to their specification.
+
+## 6. Consequences
+
+This will be a relatively big change in the types and the type checker.
+Additionally, it would render many existing specifications ill-typed, as the
+proposed solution imposes a stricter typing discipline. On the positive, the
+proposed solution is backwards-compatible with TLC, as we are proposing the
+default untyped implementation for the operators.
```

Comment:
Note that the main point of this section is to provide a space for us to evaluate the consequences of the decision *after* we have actually seen the effects. :)